patterns/structural/proxy.py

```diff
@@ -1,55 +1,88 @@
+Add functionality or logic (e.g. logging, caching, authorization) to a resource
+without changing its interface.
 """
 
-import time
 
+class Subject:
+    """
+    As mentioned in the document, interfaces of both RealSubject and Proxy should
+    be the same, because the client should be able to use RealSubject or Proxy with
+    no code change.
+
+    Not all times this interface is necessary. The point is the client should be
+    able to use RealSubject or Proxy interchangeably with no change in code.
+    """
+
+    def do_the_job(self, user):
+        raise NotImplementedError()
 
-class SalesManager:
-    def talk(self):
-        print("Sales Manager ready to talk")
 
+class RealSubject(Subject):
+    """
+    This is the main job doer. External services like payment gateways can be a
+    good example.
+    """
 
-class Proxy:
+    def do_the_job(self, user):
+        print(f'> I am doing the job for {user}')
+
+
+class Proxy(Subject):
     def __init__(self):
-        self.busy = 'No'
-        self.sales = None
-
-    def talk(self):
-        print("Proxy checking for Sales Manager availability")
-        if self.busy == 'No':
-            self.sales = SalesManager()
-            time.sleep(0.1)
-            self.sales.talk()
-        else:
-            time.sleep(0.1)
-            print("Sales Manager is busy")
+        self._real_subject = RealSubject()
+
+    def do_the_job(self, user):
+        """
+        logging and controlling access are some examples of proxy usages.
+        """
 
+        print(f'[log] Doing the job for {user} is requested.')
+
+        if user == 'admin':
+            self._real_subject.do_the_job(user)
+        else:
+            print(f'[log] I can do the job just for `admins`.')
 
-class NoTalkProxy(Proxy):
-    def talk(self):
-        print("Proxy checking for Sales Manager availability")
-        time.sleep(0.1)
-        print("This Sales Manager will not talk to you", "whether he/she is busy or not")
 
+def client(job_doer, user):
+    job_doer.do_the_job(user)
 
 if __name__ == '__main__':
-    p = Proxy()
-    p.talk()
-    p.busy = 'Yes'
-    p.talk()
-    p = NoTalkProxy()
-    p.talk()
-    p.busy = 'Yes'
-    p.talk()
-
-### OUTPUT ###
-# Proxy checking for Sales Manager availability
-# Sales Manager ready to talk
-# Proxy checking for Sales Manager availability
-# Sales Manager is busy
-# Proxy checking for Sales Manager availability
-# This Sales Manager will not talk to you whether he/she is busy or not
-# Proxy checking for Sales Manager availability
-# This Sales Manager will not talk to you whether he/she is busy or not
+    proxy = Proxy()
+    real_subject = RealSubject()
+
+    print('# doing the job with proxy:')
+    client(proxy, 'admin')
+    client(proxy, 'anonymous')
+
+    print()
+
+    print('# doing the job with real-subject:')
+    client(real_subject, 'admin')
+    client(real_subject, 'anonymous')
+
+
+OUTPUT = """
```

Comment:
I would like you to try this kind of outputs testing
https://github.com/faif/python-patterns/blob/8dbfbd67b8c178b23b84e74302563a0f6ec95cfa/patterns/behavioral/template.py#L49-L72

It is an ongoing task to move all OUTPUTs to be like doctests.